src/transformers/trainer.py

```diff
@@ -510,6 +516,10 @@ def __init__(
         ):
             self.place_model_on_device = False
 
+        self.tokenizer = processor if processor is not None else tokenizer
+        if processor is not None and hasattr(processor, "feature_extractor"):
+            tokenizer = processor.feature_extractor
```

Comment:
Note that all this is doing is setting the padding method in the default data collator:
https://github.com/huggingface/transformers/blob/15c74a28294fe9082b81b24efe58df16fed79a9e/src/transformers/trainer.py#L513-L517

There's no `pad` method defined for processors, so the processor **cannot** be used here. Only feature extractors and tokenizers have a `pad` method defined, so they are the only two viable options.

Reply:
Note here that we're setting:
```python
tokenizer = processor.feature_extractor
```
Not:
```python
self.tokenizer = processor.feature_extractor
```
=> the feature extractor is **strictly** used for padding purposes in the data collator, and not set to an attribute of the trainer

In fact, since we set:
```python
self.tokenizer = processor
```
the processor is correctly the attribute which is both saved and pushed to the hub.

Reply:
Agree though that this behaviour is somewhat "silent' to the user and can be improved on (will iterate on this once we have a design established)